src/Autonomous.cpp

```diff
@@ -243,23 +264,99 @@ void Autonomous::autonomyIter()
 		{
 			point_t rightPostLandmark = landmarks[target.right_post_id];
 			// transform and add the new data to the filter
 			point_t landmarkMapSpace = invTransform * rightPostLandmark;
 			// the filtering is done on the target in map space to reduce any phase lag
 			// caused by filtering
-			driveTarget.topRows(2) = landmarkFilter.get(landmarkMapSpace).topRows(2);
-			// clear the filter so the second post is removed from it
-			landmarkFilter.reset();
-		}
-		// If we are in a search pattern, set the drive target to the next search point
-		else if (state == NavState::SEARCH_PATTERN)
-		{
-			driveTarget = search_target;
+			point_t post_2 = gate_filter.get(landmarkMapSpace);
+
+			// we haven't seen the first post yet, use the second post as the drive target
+			if (landmarkFilter.getSize() == 0)
+			{
+				driveTarget.topRows(2) = post_2.topRows(2);
+			}
+			// we have already seen the first post
+			else
+			{
+				point_t post_1 = landmarkFilter.get();
+				point_t center = (post_1 + post_2) / 2;
+
+				// Use the center as the drive target to avoid being close to one post but far from the other
+				driveTarget.topRows(2) = center.topRows(2);
+
+				// True if we are close enough to have sufficiently accurate landmark measurements
+				bool inRange = std::isnan(gate_targets.first(2)) && dist(pose, center, 0.0) < 8.0;
+				// True if we are within refining distance
+				bool shouldRefine = !std::isnan(gate_targets.first(2)) && std::isnan(gate_targets.second(2)) && dist(pose, gate_targets.first, 0.0) < 1.5;
+
+				// we have not set gate targets yet or have potentially inaccurate gate landmark measurements
+				if (inRange || shouldRefine)
+				{
+					point_t offset{(post_1(1) - post_2(1)) / 2, (post_2(0) - post_1(0)) / 2, 0};
+					pose_t goal_1 = center + 2 * offset;
+					pose_t goal_2 = center - 2 * offset;
+
+					// Choose the closest goal to be the first target
+					if (dist(pose, goal_1, 0) < dist(pose, goal_2, 0))
+					{
+						gate_targets = std::make_pair(goal_1, goal_2);
+					}
+					else
+					{
+						gate_targets = std::make_pair(goal_2, goal_1);
+					}
+
+					// Angle between targets
+					double angle = std::atan2(gate_targets.second(1) - gate_targets.first(1), gate_targets.second(0) - gate_targets.first(0));
+					// Angle between current post and first target
+					gate_targets.first(2) = std::atan2(gate_targets.first(1) - pose(1), gate_targets.first(0) - pose(0));
+					// Mark targets as unrefined if not refining
+					gate_targets.second(2) = shouldRefine ? angle : NAN;
+
+					// Replan
+					plan_cost = INFINITE_COST;
+
+					// Align to target to improve planning
+					state = NavState::GATE_ALIGN;
+				}
+
+				// we have already seen the first post and have set gate targets
+				if (!std::isnan(gate_targets.first(2)))
+				{
+					// Update gate targets
+					if (!std::isinf(gate_targets.first(2)) &&
+						dist(pose, gate_targets.first, 0.0) < 0.2)
```

Comment:
Here you should probably use a nonzero theta weight (e.g. `dist(blah, blah, 1.0)`). We want to make sure that we've actually achieved the target orientation as well. Otherwise who knows what will happen while the robot turns to face the second target---it might crash into a gatepost.

Reply:
Actually, here's an idea. There is some duplicated code here in the sense that we often want to check how close we are to the target, and if so, advance the navigation state and re-plan. Maybe we can pull all that into a single section of code to make sure that we're doing these things consistently.

To facilitate that, we might want to clarify how we're using various local variables such as `driveTarget`. We should probably use `driveTarget` just for local control targets, and introduce a different local variable `planTarget` for the pose we're planning to reach. (Probably also a third local variable `navigationTarget` for the URC GPS target, because that's often different from the planning target, e.g. during search patterns.)

That simplifies some of the logic here, because we just need to test whether the current pose is close to the `planTarget`. If so, we should advance to the next state, compute a new `planTarget`, and trigger a replan by setting `plan_cost` to `INFINITE_COST`. The navigation states could be for instance, FOLLOWING_GPS, SEARCH_PATTERN, POST_VISIBLE, GATE_ALIGN, GATE_TRAVERSE or something like that.